connectors/source.py

```diff
@@ -549,6 +584,10 @@ def _serialize(value):
 
         return doc
 
+    def sync_cursor(self):
+        """Returns the sync cursor of the current sync"""
+        return self._sync_cursor
```

Comment:
What's the expectation on this method? Will the framework peridically check this field and store the cursor? Will it only do it once sync ended successfully/unsuccessfully?

Reply:
>Will the framework peridically check this field and store the cursor?

Not for now. Currently, we do store the `sync_cursor` in sync job when claiming it:
https://github.com/elastic/connectors-python/blob/1f4447bf95312f8aa9a39fd2c01ae73e570398eb/connectors/protocol/connectors.py#L273

But that is mainly for record purposes, so we know an incremental sync job is executed with this cursor.

This is needed when later we want to be able to `resume` a suspended sync job.

>What's the expectation on this method?
Will it only do it once sync ended successfully/unsuccessfully?

So the expectation for now is to return the cursor at the end of the sync (i.e. at the end of `get_docs`).

The sync_cursor will only be written back to `connector` after a successful sync. See discussion here: https://github.com/elastic/connectors-python/pull/963#discussion_r1209077066

Reply:
What do you think if we allow connector to set cursor instead, e.g:

```python
def get_docs(self, filtering):
    for page in self.client.get_remote_docs_page():
        for doc in page['docs']:
            yield emit_document(doc), None
       yield emit_cursor(page['cursor']
```

One important thing to keep in mind: some 3rd-party systems have cursors that expire. For example, binlog is limited in size and eventually moves on, so if you're too late, you can't do incremental any more. Same for Sharepoint Online - if your delta cursor is too old, it cannot be used any more.

So it's important even for the failed/stuck sync to move the delta on when possible. I haven't thought too deep into it, but I think this is a super important thing to address.

Reply:
>One important thing to keep in mind: some 3rd-party systems have cursors that expire. For example, binlog is limited in size and eventually moves on, so if you're too late, you can't do incremental any more. Same for Sharepoint Online - if your delta cursor is too old, it cannot be used any more.

Not sure about binlog, but the `deltaLink` for SPO is valid for 7 days. This was considered when designing incremental sync, and the conclusion is to fail the sync job if the cursor expires.

>I haven't thought too deep into it, but I think this is a super important thing to address.

Same, as I mentioned in https://github.com/elastic/connectors-python/pull/963#discussion_r1209077066, I chose the safest and easiest approach. 

>So it's important even for the failed/stuck sync to move the delta on when possible

This should work 99% of the time. The other 1% is due to the async data flow
https://github.com/elastic/connectors-python/blob/15afd8d63000d21e5f809f2af3b492928ae0a77e/connectors/es/sink.py#L14-L17

Data being extracted from a remote source doesn't mean it's ingested into ES. With the original implicit incremental sync design (i.e. run the first sync job as full sync, and incremental sync for all the subsequent sync jobs), we couldn't afford any inconsistency. 

With the current design, I think it's an acceptable edge case. WDYT? cc @tarekziade @vidok

Reply:
Thanks for the explanations, I see that the issue is on your radar and being taken care of. I think it's good to move ahead with the taken approach